File: HospitalApp/src/main/java/QRCodePane.java
import config.Configuration;
import general.security.SecurityHelper;
import general.utility.GeneralHelper;
import general.utility.Helper;
import javafx.geometry.Pos;
import javafx.scene.Group;
import javafx.scene.Scene;
import javafx.scene.control.*;
import javafx.scene.image.ImageView;
import javafx.scene.layout.HBox;
import javafx.scene.layout.VBox;
import viewmodel.Config;

import javax.crypto.SecretKey;
public class QRCodePane {
    public static Scene scene = new Scene(new Group(), Config.WIDTH, Config.HEIGHT);
    private static ImageView QRview = new ImageView();
    private static final VBox container = new VBox(50);
    private static final Button backButton = new Button("Back");
    private static final Button scanButton = new Button("Scan Signature");
    private static String record = null;

    public void setRecord (String input) {
        record = input;
    }

    public void drawQRcode(SecretKey key, long timestamp){
        byte[] Medical_Identifier = null;
        byte[] encrypted_data = null;
        byte[] hash_Output = null;
        try {
            encrypted_data = Helper.AESencrypt(record.getBytes(), key);
            Medical_Identifier = GlobalVar.fullNodeRestClient.getMedicalOrgIdentifier();
            byte[] timestamp_byte = GeneralHelper.longToBytes(timestamp);
            byte[] combined_bytes = new byte[timestamp_byte.length + encrypted_data.length + Medical_Identifier.length];

            System.arraycopy(timestamp_byte, 0, combined_bytes, 0, timestamp_byte.length);
            System.arraycopy(encrypted_data, 0, combined_bytes, timestamp_byte.length, encrypted_data.length);
            System.arraycopy(Medical_Identifier, 0, combined_bytes, timestamp_byte.length + encrypted_data.length, Medical_Identifier.length);

            hash_Output = SecurityHelper.hash(combined_bytes, Configuration.BLOCKCHAIN_HASH_ALGORITHM);

            String QRoutput = GeneralHelper.bytesToStringHex(hash_Output);
            KTHelper.drawQRCode(QRview, QRoutput);
        } catch (Exception e){e.printStackTrace();}
    }


    public void init() {
        connectComponents();
        styleComponents();
        setCallbacks();
    }

    private void connectComponents(){
        HBox tempBox = new HBox(30, backButton, scanButton);
        tempBox.setAlignment(Pos.CENTER);
        container.getChildren().addAll(QRview, tempBox);
        container.setAlignment(Pos.CENTER);
        scene.setRoot(container);
    }

    private void styleComponents(){
    }

    private void setCallbacks(){
        backButton.setOnAction(event -> {
            SceneManager.showMainMenuScene();
        });

        scanButton.setOnAction(event -> {
            SceneManager.showScanScene(ScanPane.TYPE.SIGNATURE);
        });
    }
}

File: ValidatorClient/src/main/kotlin/viewmodel/panes/QRCodePane.kt
package viewmodel.panes

import javafx.geometry.Pos
import javafx.scene.control.Button
import javafx.scene.image.ImageView
import javafx.scene.layout.BorderPane
import javafx.scene.layout.HBox
import javafx.scene.layout.VBox
import main.Helper
import viewmodel.Config
import viewmodel.SceneManager

object QRCodePane : BorderPane() {
    private val container = VBox(50.0)
    private val qrView = ImageView()
    private val backButton = Button("Back")
    private val scanButton = Button("Scan Signature")
    private val bottomBar = HBox(30.0)

    init {
        qrView.fitWidth = Config.WIDTH * 0.2
        qrView.fitHeight = Config.WIDTH * 0.2
        bottomBar.alignment = Pos.CENTER
        connectComponents()
        styleComponents()
        setCallbacks()
    }

    fun drawQRCode(str: String) {
        Helper.drawQRCode(qrView, str)
        qrView.visibleProperty().set(true)
    }

    fun removeQRCode() {
        qrView.visibleProperty().set(false)
    }

    fun hideButtons() {
        if(bottomBar.children.contains(scanButton))
            bottomBar.children.remove(scanButton)
        if(bottomBar.children.contains(backButton))
            bottomBar.children.remove(backButton)
    }

    fun showButtons() {
        if(!bottomBar.children.contains(backButton))
            bottomBar.children.add(backButton)
        if(!bottomBar.children.contains(scanButton))
            bottomBar.children.add(scanButton)
    }

    private fun connectComponents() {
        bottomBar.children.addAll(backButton, scanButton)
        container.children.addAll(
            qrView, bottomBar
        )
        container.alignment = Pos.CENTER
        center = container
    }

    private fun styleComponents() {

    }

    private fun setCallbacks() {
        backButton.setOnAction {
            SceneManager.showMainMenuScene()
        }

        scanButton.setOnAction {
            SceneManager.showScanScene(ScanPane.TYPE.SIGNATURE)
        }
    }
}
